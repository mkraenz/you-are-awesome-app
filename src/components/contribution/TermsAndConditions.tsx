import React, { FC } from "react";
import { useTranslation } from "react-i18next";
import { Linking, StyleSheet, View } from "react-native";
import { Checkbox, List, Paragraph, useTheme } from "react-native-paper";

const styles = StyleSheet.create({
    link: { color: "blue", textDecorationLine: "underline" },
    checkboxContainer: {
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
    },
});

interface Props {
    accepted: boolean;
    handlePress: () => void;
    error: boolean;
}

// TODO #229 tests
const TermsAndConditions: FC<Props> = ({ accepted, handlePress, error }) => {
    const { colors } = useTheme();
    const { t } = useTranslation();

    const renderTitle = (
        <>
            <Paragraph style={error ? { color: colors.error } : {}}>
                {t("contributeAgreeToConditions")}
            </Paragraph>
            <Paragraph
                onPress={() => Linking.openURL(t("contributeConditionsLink"))}
                style={styles.link}
            >
                {t("contributeConditions")}
            </Paragraph>
        </>
    );
    const renderCheckbox = () => (
        // in case of multiline title, center the checkbox
        <View style={styles.checkboxContainer}>
            <Checkbox.Android
                accessibilityStates={{}}
                status={accepted ? "checked" : "unchecked"}
                onPress={handlePress}
                color={colors.primary}
                uncheckedColor={error ? colors.error : ""}
            />
        </View>
    );

    return (
        <List.Item
            accessibilityStates={{}}
            title={renderTitle}
            left={renderCheckbox}
            titleNumberOfLines={2}
        ></List.Item>
    );
};

export default TermsAndConditions;
