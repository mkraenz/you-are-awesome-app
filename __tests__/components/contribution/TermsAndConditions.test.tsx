import { noop } from "lodash";
import React from "react";
import "react-native";
// Note: test renderer must be required after react-native.
import renderer, { act } from "react-test-renderer";
import TermsAndConditions from "../../../src/components/contribution/TermsAndConditions";
import LocalizedMockPaperProvider from "../../helpers/LocalizedMockPaperProvider";

it("renders correctly", async () => {
    const tree = renderer
        .create(
            <LocalizedMockPaperProvider>
                <TermsAndConditions accepted error={false} handlePress={noop} />
            </LocalizedMockPaperProvider>
        )
        .toJSON();
    await act(async () => {});

    expect(tree).toMatchSnapshot();
    expect(JSON.stringify(tree)).not.toContain(""); // empty checkbox
});

it("renders correctly on error", async () => {
    const tree = renderer
        .create(
            <LocalizedMockPaperProvider>
                <TermsAndConditions accepted={false} error handlePress={noop} />
            </LocalizedMockPaperProvider>
        )
        .toJSON();
    await act(async () => {});

    expect(tree).toMatchSnapshot();
    expect(JSON.stringify(tree)).toContain(""); // empty checkbox
});
