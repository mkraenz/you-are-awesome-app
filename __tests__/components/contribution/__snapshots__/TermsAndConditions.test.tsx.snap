// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "padding": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "center",
          }
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
                "height": 36,
                "padding": 6,
                "width": 36,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                ],
              }
            }
          >
            <Text />
            <View
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "borderColor": "#6200ee",
                    "borderWidth": 0,
                    "height": 14,
                    "width": 14,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginVertical": 6,
              "paddingLeft": 8,
            },
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={3}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {},
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {},
                  ],
                ],
              ]
            }
          >
            I have read and agree to the 
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                ],
              ]
            }
          >
            Terms & Conditions
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {},
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {},
                  ],
                ],
              ]
            }
          >
             and 
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                ],
              ]
            }
          >
            Privacy Policy
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly on error 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "padding": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "center",
          }
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
                "height": 36,
                "padding": 6,
                "width": 36,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                ],
              }
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              style={
                Array [
                  Object {
                    "color": "#B00020",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "material-community",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
            <View
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "borderColor": "#B00020",
                    "borderWidth": 0,
                    "height": 14,
                    "width": 14,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginVertical": 6,
              "paddingLeft": 8,
            },
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={3}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "#B00020",
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "#B00020",
                    },
                  ],
                ],
              ]
            }
          >
            I have read and agree to the 
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                ],
              ]
            }
          >
            Terms & Conditions
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "#B00020",
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "#B00020",
                    },
                  ],
                ],
              ]
            }
          >
             and 
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "color": "blue",
                      "textDecorationLine": "underline",
                    },
                  ],
                ],
              ]
            }
          >
            Privacy Policy
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
